streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -2616,134 +2616,107 @@ <VT, VR> KStream<K, VR> leftJoin(final KTable<K, VT> table,
-     * {@code keyValueMapper.map(stream.keyValue) == table.key}.
+     * {@code keyValueMapper.map(streamRecord) == tableRecord.key}.
      * "Table lookup join" means, that results are only computed if {@code KStream} records are processed.
-     * This is done by performing a lookup for matching records in the <em>current</em> internal {@link GlobalKTable}
-     * state.
+     * This is done by performing a lookup for matching records in the <em>current</em> (i.e., processing time)
+     * internal {@link GlobalKTable} state.
      * In contrast, processing {@link GlobalKTable} input records will only update the internal {@link GlobalKTable}
      * state and will not produce any result records.
-     * <p>
-     * For each {@code KStream} record that finds a corresponding record in {@link GlobalKTable} the provided
+     *
+     * <p>For each {@code KStream} record that finds a joining record in the {@link GlobalKTable} the provided
      * {@link ValueJoiner} will be called to compute a value (with arbitrary type) for the result record.
-     * The key of the result record is the same as the key of this {@code KStream}.
-     * If a {@code KStream} input value is {@code null} the record will not be included in the join operation
-     * and thus no output record will be added to the resulting {@code KStream}.
+     * The key of the result record is the same as the stream record's key.
+     * If you need read access to the {@code KStream} key, use {@link #join(GlobalKTable, KeyValueMapper, ValueJoinerWithKey)}.
+     * If a {@code KStream} input record's value is {@code null} or if the provided {@link KeyValueMapper keySelector}
+     * returns {@code null}, the input record will be dropped, and no join computation is triggered.
+     * If a {@link GlobalKTable} input record's key is {@code null} the input record will be dropped, and the table
+     * state won't be updated.
+     * {@link GlobalKTable} input records with {@code null} values are considered deletes (so-called tombstone) for
+     * the table.
+     *
+     * <p>Example, using the first value attribute as join key:
+     * <table border='1'>
+     * <tr>
+     * <th>KStream</th>
+     * <th>GlobalKTable</th>
+     * <th>state</th>
+     * <th>result</th>
+     * </tr>
+     * <tr>
+     * <td>&lt;K1:(GK1,A)&gt;</td>
+     * <td></td>
+     * <td></td>
+     * <td></td>
+     * </tr>
+     * <tr>
+     * <td></td>
+     * <td>&lt;GK1:b&gt;</td>
+     * <td>&lt;GK1:b&gt;</td>
+     * <td></td>
+     * </tr>
+     * <tr>
+     * <td>&lt;K1:(GK1,C)&gt;</td>
+     * <td></td>
+     * <td>&lt;GK1:b&gt;</td>
+     * <td>&lt;K1:ValueJoiner((GK1,C),b)&gt;</td>
+     * </tr>
+     * </table>
+     *
+     * In contrast to {@link #join(KTable, ValueJoiner)}, there is no co-partitioning requirement between this
+     * {@code KStream} and the {@link GlobalKTable}.
+     * Also note, that the {@link GlobalKTable} is updated "asynchronously", and thus this operation is inherently
```

Comment:
I wonder if you can make this more precise (I cannot, btw). Essentially, there are no guarantees on the ordering of the join w.r.t. stream time, right?

Reply:
Yeah, maybe just expand on it a tiny bit. But I think you are right, maybe that is best explained in the actual doc. I wasn't aware that there is a section on this in the documentation.

Maybe I was just triggered by the quotation marks. You could consider something like this: `Also note that there are no ordering guarantees between the updates on the left and the right side of this join, since updates to the {@link GlobalKTable} are in no way synchronized to this join. Therefore, the result of the join is inherently non-deterministic`.

But if you don't like the proposal, feel free to just keep it as is.

Reply:
> I wasn't aware that there is a section on this in the documentation.

To be fair, not sure to what extend we actually cover this in the docs. But if we want to document it better, let's do it in the docs, not JavaDocs.

> Maybe I was just triggered by the quotation marks.

Yeah. In the end, it's hard to describe and I was trying to use single word, but asynchronously does not fully fit (it's much more nuanced) so I used quotation marks to indicate it's not a totally accurate description.

I like your suggestions.